### PR TITLE
`UTF8::computeWidth` skip ANSI escape sequences

### LINKED ISSUE #59020: The `UTF8::computeWidth` function should skip ANSI escape sequences
**Use case**

See the picture. There is one unpleasant detail: the width of a field is calculated incorrectly:

Screenshot_20240120_105734

**Describe the solution you'd like**

Ignore the symbols inside https://en.wikipedia.org/wiki/ANSI_escape_code#CSIsection in width calculation.

Test:

```
:) SELECT format('\x1b[38;2;{0};{1};{2}m█\x1b[0m', 255, 128, 0) AS x

┌─x────────────────────┐
│ █ │
└──────────────────────┘
```

Should be:

```
:) SELECT format('\x1b[38;2;{0};{1};{2}m█\x1b[0m', 255, 128, 0) AS x

┌─x─┐
│ █ │
└───┘
```

PS. The original query is

`SELECT id, max(score) AS s, max(descendants), argMax(title, update_time), arrayStringConcat(arrayMap(x -> format('\x1b[38;2;{0};{1};{2}m█\x1b[0m', least(greatest(sqrt(x / 10) * 255, 0), 255)::UInt8::String, least(greatest(sqrt((x - 10) / 50) * 255, 0), 255)::UInt8::String, least(greatest(sqrt((x - 50) / 250) * 255, 0), 255)::UInt8::String), arrayDifference(arrayFill(x -> x != 0, maxResample(now() - INTERVAL 2 DAY, now(), 3600)(score, update_time))))) AS bar FROM hackernews_changes_items WHERE id IN (SELECT arrayJoin(ids) FROM (SELECT ids FROM hackernews_top WHERE type = 'top' ORDER BY update_time DESC LIMIT 1)) GROUP BY id ORDER BY s DESC`

### PR DESCRIPTION
**Pull request by Blargian**
PR introduces changes such that symbols inside [ANSI escapes](https://en.wikipedia.org/wiki/ANSI_escape_code#CSIsection) are ignored in the width calculation.

Closes [#59020](https://github.com/ClickHouse/ClickHouse/issues/59020)

### Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Width of fields now correctly calculate, ignoring ANSI escape sequences.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit c2dd365e3724aff2837ce70509b714194ffe1577 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63270/c2dd365e3724aff2837ce70509b714194ffe1577/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Common/UTF8Helpers.cpp [hunk 1]
```diff
@@ -97,13 +97,14 @@ namespace
 enum ComputeWidthMode
 {
     Width,              /// Calculate and return visible width
-    BytesBeforLimit     /// Calculate and return the maximum number of bytes when substring fits in visible width.
+    BytesBeforeLimit     /// Calculate and return the maximum number of bytes when substring fits in visible width.
```
Comment (antaljanosbenjamin) [inline src/Common/UTF8Helpers.cpp:100]:
```suggestion
    BytesBeforeLimit    /// Calculate and return the maximum number of bytes when substring fits in visible width.
```

File: src/Common/UTF8Helpers.cpp [hunk 2]
```diff
@@ -162,20 +174,18 @@ size_t computeWidthImpl(const UInt8 * data, size_t size, size_t prefix, size_t l
...
-                // (`0x08`, `0x09`, `0x0a`, `0x0b`, `0x0c`, `0x0d`, `0x1b`)
-                // Since we don't touch the original column data, there is no easy way to escape them.
-                // TODO: escape control characters
                 // TODO: multiline support for '\n'
 
-                // special treatment for '\t'
+                // special treatment for '\t' and for ESC
                 size_t next_width = width;
+                if (decoder.codepoint == '\x1b')
+                    isEscapeSequence = true;
```
Comment (antaljanosbenjamin) [inline src/Common/UTF8Helpers.cpp:182]: Let's turn the next `if` into `else if`, it is a trivial change.

File: src/Common/UTF8Helpers.cpp [hunk 3]
```diff
@@ -132,21 +133,32 @@ size_t computeWidthImpl(const UInt8 * data, size_t size, size_t prefix, size_t l
...
         }
 #endif
 
         while (i < size && isPrintableASCII(data[i]))
         {
-            ++width;
+            if (!isEscapeSequence)
+                ++width;
+            if (isCSIFinalByte(data[i]) && data[i-1]!='\x1b')
+                isEscapeSequence = false; /// end of CSI escape sequence reached
```
Comment (antaljanosbenjamin) [inline src/Common/UTF8Helpers.cpp:154]:
```suggestion
            else if (isCSIFinalByte(data[i]) && data[i-1]!='\x1b')
                isEscapeSequence = false; /// end of CSI escape sequence reached
```

File: tests/queries/0_stateless/03142_skip_ANSI_in_UTF8_compute_width.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+SELECT format('\x1b[38;2;{0};{1};{2}m█\x1b[0m', 255, 128, 0) AS x FORMAT Pretty;
```
Comment (antaljanosbenjamin) [inline tests/queries/0_stateless/03142_skip_ANSI_in_UTF8_compute_width.sql:1]:
```suggestion
SELECT format('\x1b[38;2;{0};{1};{2}m█\x1b[0m', 255, 128, 0) AS x FORMAT Pretty;

```

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]

File: tests/queries/0_stateless/03142_skip_ANSI_in_UTF8_compute_width.sql (hunk 4 above)
Comment (Blargian) [reply tests/queries/0_stateless/03142_skip_ANSI_in_UTF8_compute_width.sql]: @antaljanosbenjamin the change looks identical, apart from an empty line after the query - just curious if there is some specific reason for this?

**Review by alexey-milovidov — APPROVED** [review]

File: tests/queries/0_stateless/03142_skip_ANSI_in_UTF8_compute_width.sql (hunk 4 above)
Comment (antaljanosbenjamin) [reply tests/queries/0_stateless/03142_skip_ANSI_in_UTF8_compute_width.sql]: It is a [good practice](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline) in general to have every text file end with an empty line.

**Review by antaljanosbenjamin — APPROVED** [review]

File: src/Common/StringUtils/StringUtils.h [hunk 5]
```diff
@@ -138,6 +138,12 @@ inline bool isPrintableASCII(char c)
     return uc >= 32 && uc <= 126;   /// 127 is ASCII DEL.
 }
 
+inline bool isCSIFinalByte(char c)
```
Comment (rschu1ze) [inline src/Common/StringUtils/StringUtils.h:141]: \+ `/// Control Sequence Introducer, https://en.wikipedia.org/wiki/ANSI_escape_code#CSIsection` for simple-minded people like me who don't know what "CSI" means

**Comment by alexey-milovidov** [comment]
@rschu1ze @Blargian, it does not work:
Screenshot_20240530_163637

**Comment by Blargian** [comment]
I'll take a look

**Comment by alexey-milovidov** [comment]
@Blargian this is the reproducer:

```
milovidov-pc :) SELECT 'Hello', format('\x1b[38;2;{0};{1};{2}m█\x1b[0m test \x1b[38;2;{0};{1};{2}m█\x1b[0m', 255, 128, 128) AS x

SELECT
    'Hello',
    format('0};{1};{2}m█ test 0};{1};{2}m█', 255, 128, 128) AS x

Query id: 33bd2f4e-bddc-47d8-9ce5-a310a2e4cfed

   ┌─'Hello'─┬─x───────────┐
1. │ Hello   │ █ test █ │
   └─────────┴─────────────┘
```

**Comment by alexey-milovidov** [comment]
@Blargian this is also wrong:

```
SELECT visibleWidth('0};{1};{2}m█')

Query id: 650f1184-8606-45c9-9fed-d44685e9013e

   ┌─visibleWidth('0};{1};{2}m█')─┐
1. │                           12 │
   └──────────────────────────────┘
```

**Comment by Blargian** [comment]
> @Blargian this is also wrong:
> 
> ```
> SELECT visibleWidth('0};{1};{2}m█')
> 
> Query id: 650f1184-8606-45c9-9fed-d44685e9013e
> 
>    ┌─visibleWidth('0};{1};{2}m█')─┐
> 1. │                           12 │
>    └──────────────────────────────┘
> ```

Thanks, I'm taking a look now - sorry for the delay.

**Comment by alexey-milovidov** [comment]
@Blargian sorry, it is just formatted incorrectly, but it should be: two 24-bit color escape sequences.

**Comment by alexey-milovidov** [comment]
```
SELECT visibleWidth('\x1b[38;2;{0};{1};{2}m█\x1b[0m')
```

**Pull request merged**